packages/components/src/number-control/index.tsx

```diff
@@ -25,16 +25,20 @@ import type { NumberControlProps } from './types';
 import { HStack } from '../h-stack';
 import { Spacer } from '../spacer';
 import { useCx } from '../utils';
+import { useDeprecated36pxDefaultSizeProp } from '../utils/use-deprecated-props';
 
 const noop = () => {};
 
 function UnforwardedNumberControl(
-	{
+	props: WordPressComponentProps< NumberControlProps, 'input', false >,
+	forwardedRef: ForwardedRef< any >
+) {
+	const {
 		__unstableStateReducer: stateReducerProp,
 		className,
 		dragDirection = 'n',
 		hideHTMLArrows = false,
-		spinControls = 'native',
+		spinControls = hideHTMLArrows ? 'none' : 'native',
```

Comment:
This is now a `const` and immutable, so I streamlined the code a bit. The conditional logic itself should be the same as before.

Reply:
If `hideHTMLArrows` changes during the component's lifetime, would the value of `spinControls` change too, or would it stay the same as on the first render?

If that's the case, the change to the logic proposed in this PR may introduce a small regression.

Reply:
I just did some testing on the impact of `hideHTMLarrows` changing during the component's lifetime, and it does also change the value of `spinControls`, which is consistent with `trunk`.

Does that address the concern you had here @ciampo? I want to make sure I'm reading your comment correctly.

<details>
<summary>Here's a diff of my testing story, in case you want to check it out.</summary>

```diff
diff --git a/packages/components/src/number-control/index.tsx b/packages/components/src/number-control/index.tsx
index e0a09ec06a..7b14685172 100644
--- a/packages/components/src/number-control/index.tsx
+++ b/packages/components/src/number-control/index.tsx
@@ -100,7 +100,7 @@ function UnforwardedNumberControl(
 		}
 		return constrainValue( nextValue, shift ? delta : undefined );
 	};
-
+	console.log( { hideHTMLArrows, spinControls } );
 	/**
 	 * "Middleware" function that intercepts updates from InputControl.
 	 * This allows us to tap into actions to transform the (next) state for
diff --git a/packages/components/src/number-control/stories/index.story.tsx b/packages/components/src/number-control/stories/index.story.tsx
index 3588063f0f..5448e33198 100644
--- a/packages/components/src/number-control/stories/index.story.tsx
+++ b/packages/components/src/number-control/stories/index.story.tsx
@@ -12,6 +12,7 @@ import { useState } from '@wordpress/element';
  * Internal dependencies
  */
 import NumberControl from '..';
+import Button from '../../button';
 
 const meta: Meta< typeof NumberControl > = {
 	title: 'Components (Experimental)/NumberControl',
@@ -62,3 +63,28 @@ export const Default = Template.bind( {} );
 Default.args = {
 	label: 'Value',
 };
+
+const ChangeHTMLArrowsTemplate: StoryFn< typeof NumberControl > = ( {
+	...props
+} ) => {
+	const [ localHideHTMLArrows, setLocalHideHTMLArrows ] = useState( false );
+
+	return (
+		<>
+			<NumberControl
+				{ ...props }
+				hideHTMLArrows={ localHideHTMLArrows }
+			/>
+			<Button
+				onClick={ () =>
+					setLocalHideHTMLArrows( ! localHideHTMLArrows )
+				}
+			>
+				{ localHideHTMLArrows
+					? 'show html arrows'
+					: 'hide html arrows' }
+			</Button>
+		</>
+	);
+};
+export const ChangeHTMLArrows = ChangeHTMLArrowsTemplate.bind( {} );

```

</details>

The values of both props are logged to the console when you click the button to toggle `hideHTMLArrows`.

Reply:
Yup that's it! Thank you for confirming that everything works as expected, I admit that I had never tested this pattern before.